lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -44,6 +49,55 @@ def __init__(self, config: PbenchServerConfig, logger: logging.Logger):
             role=API_OPERATION.READ,
         )
 
+    def get_paginated_obj(
+        self, query: Query, json: JSON, url: str
+    ) -> Tuple[List, Dict[str, str]]:
+        """
+        Helper function to return a slice of datasets (constructed according to the
+        user specified limit and an offset number) and a paginated object containing next page
+        url and total items count.
+        E.g. specifying the following limit and offset values will result in the corresponding
+        dataset slice:
+        "limit": 10, "offset": 20 -> dataset[20: 30]
+        "limit": 10 -> dataset[0: 10]
+        "offset": 20 -> dataset[20: total_items_count]
+        TODO: We may need to optimize the pagination
+            e.g Use of unique pointers to record the last returned row and then use
+            this pointer in subsequent page request instead of an initial start to
+            narrow down the result.
+        """
+        paginated_result = {}
+        total_count = query.count()
+        query = query.order_by(Dataset.name)
+
+        # Get the user specified limit, otherwise return all the items
+        limit = json.get("limit", total_count + 1)
+        query = query.limit(limit)
+
+        # Shift the query search by user specified offset value,
+        # otherwise return the batch of result starting from the
+        # first queried item.
+        offset = json.get("offset", 0)
+        items = query.offset(offset).all()
+
+        n = len(items)
+        if n < limit:
+            next_url = ""
+        else:
+            json["offset"] = offset + n
+            parsed_url = urlparse(url)
+            next_url = (
+                parsed_url.scheme
+                + "://"
+                + parsed_url.netloc
+                + parsed_url.path
+                + "?"
+                + urlencode(json)
+            )
```

Comment:
Don't we want to retain the previous query parameters?

You could use `parsed_url._replace(query=urlencode(json))` to avoid picking apart all of the pieces. And you can use `urllib.parse.parse_qs` to turn the raw parsed query string into a dict which you could then `update` with your `json`. Maybe something like ...

```
    parsed_url = urlparse(url)
    next_url = parsed_url._replace(
        query=urlencode(parse_qs(parsed_url.query) | json)
    )
```

Reply:
That's annoying:  the `_replace()` method looks like a really useful widget, but the leading underscore indicates that it is internal to its defining class and not for general use; consistent with this, there is no documentation for the method in the obvious place; and, yet, the documentation for [`urlparse()`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse) mentions it and shows examples of what it does and how to use it.  😝

My apologies @dbutenhof, but could you explain what the `| json` does, there?

Reply:
Note that `|` also *replaces* keys in the left-hand dict with new values from the right-hand dict: which means merging a URL that already has a `&offset=old` will *update* it to `&offset=new`.

Reply:
hmm `_replace` is an interesting choice, I didn't come across it when I read the documentation. Thanks, @dbutenhof and @webbnh for pointing out the use of the method but note that this method is not documented for python version 3.8 and earlier so I wonder whether it is added on python version 3.9 onwards, and will this be a problem for us? 

> Don't we want to retain the previous query parameters?

@dbutenhof, the `json` that gets passed to this method is the original query parameters and we just update the offset value of the json and everything else remains the same. So we do retain all the original query parameters.

Reply:
The documentation is minimal; it's mentioned, with an example, on https://docs.python.org/3/library/urllib.parse.html#url-parsing, and it works fine on my Python 3.6 installation.

(`dict1 | dict2`, however, doesn't work, and appears to be new in 3.9; unfortunately, `dict.update()` doesn't return a value, so the code is a bit messier than the sample I'd written.)